hnn_core/network_builder.py

```diff
@@ -538,6 +547,8 @@ def aggregate_data(self):
         _PC.allreduce(self._nrn_dipoles['L2_pyramidal'], 1)
         for nrn_arr in self._nrn_rec_arrays.values():
             _PC.allreduce(nrn_arr._nrn_voltages, 1)
+        for recording_callback in self._nrn_rec_callbacks:
+            _CVODE.extra_scatter_gather_remove(recording_callback)
```

Comment:
would the segmentation fault return if you removed this line?

Reply:
It would. The `_remove` is necessary and sufficient.

Reply:
interesting. I'm guessing it didn't work when attached to the class because the class wasn't getting destroyed. 

Still weird, somehow this reminds me of the [netcon](https://github.com/jasmainak/netcon_bug) issue we encountered. Things would work when attached or not attached to a class. In that case, Robert argued that users are responsible for understanding Python garbage collection and objects going out of scope. Which I found weird. Neuron objects should have checks and errors IMO ...

Reply:
This is a great find @cjayb. Super nuanced though and I'm bamboozled by the necessity for `extra_scatter_gather_remove()` and the callback function to be separate from `_ExtracellularArrayBuilder`.

Reply:
Turns out it's probably a pointer/reference-thing. I made a little note in the PR description above.